connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginClassLoader.java

```diff
@@ -45,8 +75,13 @@ public String toString() {
         return "PluginClassLoader{pluginLocation=" + pluginLocation + "}";
     }
 
+    // This method needs to be thread-safe because it is supposed to be called by multiple
+    // Connect tasks. While findClass is thread-safe, defineClass called within loadClass of the
+    // base method is not. More on multithreaded classloaders in:
+    // https://docs.oracle.com/javase/7/docs/technotes/guides/lang/cl-mt.html
     @Override
-    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+    protected synchronized Class<?> loadClass(String name, boolean resolve)
```

Comment:
Does the heavy-handed locking with `synchronized` rather than implementing the multithreaded loading with finer-grained locking have some motivation here? It's protected API, but URLClassLoader already has some fine-grained locking via ClassLoader's `getClassLoadingLock(name)`.

I'm mainly curious because this approach is actually different from what is recommended in the "Recommendations for Multithreaded Custom Class Loaders" of the doc you linked. It seems like it should work, but disables multithreaded gains added in JDK7. The implementation used in `ClassLoader` seems to just allocate a lock object per `name` which is pretty easy to just copy even if we end up using different lock objects here than in `ClassLoader`.

Reply:
I had considered more fine-grain locking, but at the time it seemed to add complexity. 

Now it seems an awfully good idea after all. 

I'm inclined to use the available post Java 7 mechanisms for that: `registerAsParallelCapable` and `getClassLoadingLock`.